### PR TITLE
Review ethereumjs-vm examples, add run-blockchain example

### PR DESCRIPTION
**Pull request by diasdavid**
@wanderer @kumavis I've spent some time reviewing the examples and ended up adding more comes + restructure a bit to make the reading simpler. Also added another example, extracted from tests, on how to run the blockchain while creating the genesis block.

Will use this PR to also ask questions directly in the examples codebase

### CODE REVIEW

File: examples/run-blockchain/index.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+vm.on('beforeTx', function (tx) {
+  tx._homestead = true
+})
+
+vm.on('beforeBlock', function (block) {
+  block.header.isHomestead = function () {
+    return true
+  }
+})
```
Comment (diasdavid) [inline examples/run-blockchain/index.js]: I remember @wanderer saying that this is not relevant anymore. Should I remove it completely?

File: examples/run-blockchain/index.js [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+        var block = new Block(
+            new Buffer(raw.rlp.slice(2), 'hex'))
+
+        // forces the block into thinking they are homestead
+        block.header.isHomestead = function () {
+          return true
+        }
+        block.uncleHeaders.forEach(function (uncle) {
+          uncle.isHomestead = function () { return true }
+        })
```
Comment (diasdavid) [inline examples/run-blockchain/index.js]: same as https://github.com/ethereumjs/ethereumjs-vm/pull/86/files#r78901688

File: examples/run-blockchain/index.js [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      } catch (err) {
+        cb()
+      }
+    }
+  },
+
+  // runBlockchain aka verify all the blocks till date
+  function runBlockchain (next) {
+    vm.runBlockchain(next)
+  },
```
Comment (diasdavid) [inline examples/run-blockchain/index.js]: How would it behave in the case of a missing block? What info can we expect?

File: examples/run-blockchain/index.js [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+  // get the blockchain head
+  function getHead (next) {
+    vm.blockchain.getHead(function (err, block) {
+      // make sure the state is set before checking post conditions
+      state.root = block.header.stateRoot
```
Comment (diasdavid) [inline examples/run-blockchain/index.js]: Why does this have to be done manually? Shouldn't the state.root already be the latest head after running the blockchain?

File: examples/run-blockchain/index.js [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        if (testData.exec && key === testData.exec.address) {
+          testData.root = storageTrie.root
+        }
+        state.put(new Buffer(key, 'hex'), account.serialize(), function () {
+          cb2()
+        })
+      }
+    ], callback)
+  }, done)
+}
```
Comment (diasdavid) [inline examples/run-blockchain/index.js]: can I get a description of what all of this function is trying to achieve?

**Comment by diasdavid** [comment]
To confirm: from what I understand, mining is still not available in JS, right?

File: examples/run-blockchain/index.js (hunk 1 above)
Comment (wanderer) [inline examples/run-blockchain/index.js]: yep! You should be able to remove the homestead stuff.

File: examples/run-blockchain/test-data.json [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        "bloom": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
+        "coinbase": "8888f1f195afa192cfee860698584c030f4c9db1",
+        "difficulty": "0x020000",
+        "extraData": "0x",
+        "gasLimit": "0x023ec6",
+        "gasUsed": "0x021536",
+        "hash": "f53f268d23a71e85c7d6d83a9504298712b84c1a2ba220441c86eeda0bf0b6e3",
+        "mixHash": "29f07836e4e59229b3a065913afc27702642c683bba689910b2b2fd45db310d3",
+        "nonce": "8957e6d004a31802",
+        "number": "0x01",
```
Comment (wanderer) [inline examples/run-blockchain/test-data.json]: oh! so all that homestead stuff i said was good to remove... well this is why we have it. The block number is set to 1. The homestead is detected on the block number which  is much higher, yet the test expect the VM to run in homestead mode! Unfortaintly you can just change the number either, b/c it will invalidated the PoW. On option for the example could be just to change the number and don't check do the the PoW check

Comment (wanderer) [inline examples/run-blockchain/test-data.json]: The other option is to stop supporting anything before homestead. I really think this is the best option. We would need to `grep -r homestead` in ethereumjs-vm and ethereumjs-block and remove all homestead conditionasl. (there shouldn't be to many)

@axic if we start doing this do you think it would be a good time to back port tx, and block type change from ewasm?

File: examples/run-blockchain/index.js [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+      next(err)
+    })
+  }
+], function () {
+  console.log('--- Finished processing the BlockChain ---')
+  console.log('New head:', blockchain.meta.rawHead.toString('hex'))
+  console.log('Expected:', testData.lastblockhash)
+})
+
+function formatBlockHeader (data) {
```
Comment (wanderer) [inline examples/run-blockchain/index.js]: it would be nice to get rid of this. The problem this was solving was that the official tests had inconsistent formatting.

File: examples/run-blockchain/test-data.json [hunk 8]
```diff
@@ -0,0 +1,159 @@
+{
+  "blocks": [
+    {
+      "blockHeader": {
+        "bloom": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
```
Comment (wanderer) [inline examples/run-blockchain/test-data.json]: if we make sure that all the hex values are prepended with '0x' we can drop a bunch the formating stuff

File: examples/run-blockchain/index.js (hunk 3 above)
Comment (wanderer) [inline examples/run-blockchain/index.js]: It will give a validation error too the  callback `next`. We can remove this function be shorten this to just  `vm.runBlockchain,` if wanted.

File: examples/run-blockchain/index.js (hunk 4 above)
Comment (wanderer) [inline examples/run-blockchain/index.js]: this is for tests where we run chain A then chain B is revealed  and chain B turns out to be invalid or shorter. It shouldn't be need with the test below

File: examples/run-blockchain/index.js (hunk 5 above)
Comment (wanderer) [inline examples/run-blockchain/index.js]: This is  creating the state that the tests are suppose to start with. The tests contain a `pre` feild that has address -> account maps. This populate state Trie with that information. it is useing the helper class [accounts](https://github.com/ethereumjs/ethereumjs-account)

File: examples/run-blockchain/index.js [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+        async.forEachSeries(keys, function (key, cb3) {
+          var val = acctData.storage[key]
+          val = rlp.encode(new Buffer(val.slice(2), 'hex'))
+          key = utils.setLength(new Buffer(key.slice(2), 'hex'), 32)
+
+          storageTrie.put(key, val, cb3)
+        }, cb2)
+      },
+      function (cb2) {
+        account.setCode(state, codeBuf, cb2)
```
Comment (wanderer) [inline examples/run-blockchain/index.js]: this stores the contracts code hash in the state trie. We have just save the code to disk (instead of having a unified way to handle values, unfortainly)

File: examples/run-blockchain/index.js [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+    storageTrie.root = null
+
+    async.series([
+      function (cb2) {
+        var keys = Object.keys(acctData.storage)
+
+        async.forEachSeries(keys, function (key, cb3) {
+          var val = acctData.storage[key]
+          val = rlp.encode(new Buffer(val.slice(2), 'hex'))
+          key = utils.setLength(new Buffer(key.slice(2), 'hex'), 32)
```
Comment (wanderer) [inline examples/run-blockchain/index.js]: this stuff here is to deal with inconsistent formatting in the tests

File: examples/run-blockchain/index.js [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+          storageTrie.put(key, val, cb3)
```
Comment (wanderer) [inline examples/run-blockchain/index.js]: this create the merkle trie for storage for the accounts. This can be confusing i think. Our merkle trie looks like this.
```
stateTrie {
   address : account {
                         balance,
                         nonce,
                         codeHash -> code (directly in leveldb),
                         storageRoot -> storageTrie {key: vale} 
                     }
}
```

File: examples/run-blockchain/index.js [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        }, cb2)
+      },
+      function (cb2) {
+        account.setCode(state, codeBuf, cb2)
+      },
+      function (cb2) {
+        account.stateRoot = storageTrie.root
```
Comment (wanderer) [inline examples/run-blockchain/index.js]: after populating the storage trie this creates the ` storageRoot -> storageTrie {key: vale} ` link

File: examples/run-blockchain/index.js [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+
+        if (testData.exec && key === testData.exec.address) {
```
Comment (wanderer) [inline examples/run-blockchain/index.js]: this was for some old block tests.  I think it should be safe to remove of the test example

File: examples/run-blockchain/index.js [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+const BN = utils.BN
+const rlp = utils.rlp
+const testData = require('./test-data')
+// inMemory blockchainDB
+var blockchainDB = new Level('', { db: require('memdown') })
+
+var state = new Trie()
+
+var blockchain = new Blockchain(blockchainDB)
+blockchain.ethash.cacheDB = new Level('./.cachedb')
```
Comment (wanderer) [inline examples/run-blockchain/index.js]: the first time the ethash runs it has to generate a large data set that is stored in here.

File: examples/run-blockchain/index.js (hunk 7 above)
Comment (diasdavid) [inline examples/run-blockchain/index.js]: So, it would be just adding the `0x` to the hex values that missing it on the test-data.json?

File: examples/run-blockchain/test-data.json (hunk 8 above)
Comment (diasdavid) [inline examples/run-blockchain/test-data.json]: ✔️

File: examples/run-blockchain/test-data.json (hunk 6 above)
Comment (diasdavid) [inline examples/run-blockchain/test-data.json]: Any developments on this? I guess that for demo purposes having these is ok, the big question is the amount of monkey patching I'll have to make to apply on the blocks incoming from go-ethereum (and when to know how to stop patching)

Comment (wanderer) [inline examples/run-blockchain/test-data.json]: if you getting blocks from go-ethereum then this should be a problem since the blocks are correct with respect to the homestead changes. This is only done for the testing.

File: examples/run-blockchain/index.js (hunk 12 above)
Comment (diasdavid) [inline examples/run-blockchain/index.js]: what do we use it for?
